core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -507,6 +520,27 @@ private boolean canProcessDictionary(Page page)
         return true;
     }
 
+    private boolean canProcessLowCardinalityDictionary(Page page)
+    {
+        // We don't have to rely on 'optimizer.dictionary-aggregations' here since there is little to none chance of regression
+        if (channels.length <= 1) {
+            return false;
+        }
+        int positionCount = page.getPositionCount();
+        long cardinality = 1;
+        for (int channel : channels) {
+            if (!(page.getBlock(channel) instanceof DictionaryBlock)) {
+                return false;
+            }
+            cardinality = multiplyExact(cardinality, ((DictionaryBlock) page.getBlock(channel)).getDictionary().getPositionCount());
+            if (cardinality > positionCount * SMALL_DICTIONARIES_MAX_CARDINALITY_RATIO || cardinality > SMALL_DICTIONARIES_MAX_CARDINALITY) {
```

Comment:
nit: I would remove `cardinality > SMALL_DICTIONARIES_MAX_CARDINALITY` and only leave ratio unless there is strong reason to keep this check